### PR TITLE
Dinic Max Flow algorithm/Test class for Max Flow algorithms comparison

### PR DESCRIPTION
**Pull request by LightnessOfBeing**
Implemented Dinic maximum flow algorithm. Added test class for max flow algorithms comparison.

### CODE REVIEW

**Comment by jkinable** [comment]
To test the performance of various max flow algorithms, please modify this class: https://github.com/jgrapht/jgrapht/tree/master/jgrapht-core/src/test/java/org/jgrapht/perf/flow

**Comment by jkinable** [comment]
I haven't had the opportunity to do any testing, but, out of curiosity, how's Dinic performing compared to the other flow implementations? Are there any (or all) graph types where Dinic is doing better?

**Comment by LightnessOfBeing** [comment]
In general, Dinic is faster than both Edmonds Karp and Push Relabel. In directed graphs with random capacities and N <= 1000 and M <= 5000 Dinic outperforms Edmonds Karp and so I can say for undirected graphs (quite reasonable, since we have more edges). I wouldn't say advantage is huge, but it is still relevant. However in graphs with equal capacities Edmonds Karp works faster. 

On some inputs (about 40% of total runs) Push Relabel works extremely slow (way slower than both EK and Dinic) for some reason, on other inputs it's a little slower than/comparable to Edmonds Karp and Dinic.

File: jgrapht-core/src/test/java/org/jgrapht/alg/flow/DinicMFImplTest.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
+package org.jgrapht.alg.flow;
```
Comment (jsichi) [inline jgrapht-core/src/test/java/org/jgrapht/alg/flow/DinicMFImplTest.java]: Test files and classes also need the standard headers.

File: jgrapht-core/src/main/java/org/jgrapht/alg/flow/DinicMFImpl.java [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+
+    /**
+     * Extension for vertex class.
+     */
+    class VertexExtension
+        extends VertexExtensionBase
+    {
+
+        /**
+         * Pointer on the first "not deleted"(we do not delete an edge explicitly, rather moving pointer forward) edge.
```
Comment (jsichi) [inline jgrapht-core/src/main/java/org/jgrapht/alg/flow/DinicMFImpl.java]: This wording needs improvement.

**Comment by jsichi** [comment]
@jkinable do you have further comments on this one?

**Comment by jkinable** [comment]
Looks good to me, thanks!

**Comment by jkinable** [comment]
I think the Push-Relabel implementation needs some further investigation. This implementation seems to have some issues (it is not supposed to be super slow on some graphs). Might be related to issue #461

**Comment by AlexandruValeanu** [comment]
@jkinable I'll look into #461

**Pull request merged**